compiler/rustc_lint/src/unused.rs

```diff
@@ -1030,21 +1040,22 @@ impl EarlyLintPass for UnusedParens {
     }
 
     fn check_ty(&mut self, cx: &EarlyContext<'_>, ty: &ast::Ty) {
+        if let ast::TyKind::Array(_, len) = &ty.kind {
+            self.check_unused_delims_expr(
+                cx,
+                &len.value,
+                UnusedDelimsCtx::ArrayLenExpr,
+                false,
+                None,
+                None,
+            );
+        }
         if let ast::TyKind::Paren(r) = &ty.kind {
```

Comment:
can you set `self.with_self_ty_parens` to `false` at the end of this branch and instead assert that it is false in `exit_where_predicate`?

Because right now I think you don't lint for the inner paren in `(dyn Trait<(for<'a> fn(&'a ())>): Trait,` even though we should.